cpp/locality/BondHistogramCompute.h

```diff
@@ -0,0 +1,133 @@
+ * over these copies. It also offers some generalized functionality for process
+ * of accumulating histograms over many frames, assuming that computations must
+ * be performed on a per-NeighborBond basis. 
+*/
+class BondHistogramCompute
+{
+public:
+    //! Default constructor
+    BondHistogramCompute() : m_box(box::Box()), m_frame_counter(0), m_n_points(0), m_n_query_points(0), m_reduce(true), m_histogram(), m_local_histograms() {}
+
+    //! Destructor
+    virtual ~BondHistogramCompute() {};
+
+    //! Reset the RDF array to all zeros
+    virtual void reset()
+    {
+        m_local_histograms.reset();
+        m_frame_counter = 0;
+        m_reduce = true;
+    }
+
+    //! Reduce thread-local arrays onto the primary data arrays.
+    virtual void reduce() = 0;
+
+    //! Get the simulation box
+    const box::Box& getBox() const
+    {
+        return m_box;
+    }
+
+    //! Return :code:`thing_to_return` after reducing.
+    template<typename U>
+    U &reduceAndReturn(U &thing_to_return)
+    {
+        if (m_reduce == true)
+        {
+            reduce();
+        }
+        m_reduce = false;
+        return thing_to_return;
+    }
+
+    //! Get a reference to the bin counts array
+    const util::ManagedArray<unsigned int> &getBinCounts()
+    {
+        return reduceAndReturn(m_histogram.getBinCounts());
+    }
+
+    //! Return the bin centers.
+    /*! This vector will be of size axis.size() for each axis.
+     */
+    std::vector<std::vector<float> > getBinCenters() const
+    {
+        return m_histogram.getBinCenters();
+    }
+
+    //! Return the edges of bins.
+    /*! This vector will be of size axis.size()+1 for each axis.
+     */
+    std::vector<std::vector<float> > getBinEdges() const
+    {
+        return m_histogram.getBinEdges();
+    }
+
+    //! Return a vector of tuples (min, max) indicating the bounds of each axis.
+    std::vector<std::pair<float, float> > getBounds() const
+    {
+        return m_histogram.getBounds();
+    }
+
+    //! Return a vector indicating the number of bins in each axis.
+    std::vector<unsigned int> getAxisSizes() const
+    {
+        return m_histogram.getAxisSizes();
+    }
+
+    //! \internal
+    // Wrapper to do accumulation.
+    /*! \param neighbor_query NeighborQuery object to iterate over
+        \param query_points Points
+        \param n_query_points Number of query_points
+        \param nlist Neighbor List. If not NULL, loop over it. Otherwise, use neighbor_query
+           appropriately with given qargs.
+        \param qargs Query arguments
+        \param cf An object with operator(NeighborBond) as input.
+    */
+    template<typename Func>
+    void accumulateGeneral(const locality::NeighborQuery* neighbor_query, 
+                           const vec3<float>* query_points, unsigned int n_query_points,
+                           const locality::NeighborList* nlist,
+                           freud::locality::QueryArgs qargs,
```

Comment:
We're very inconsistent about namespaces throughout freud -- lots of places specify `freud::locality::something`. Above, it just uses `locality::NeighborQuery`, but that's very uncommon in the source. This is just a reminder in case we decide that standardization is important.

Reply:
Namespaces are one area where I'm not concerned with standardization tbh. I don't think there's necessarily a clear best practice or need to refer to the least qualified namespace. In general I try to use only the required amount of qualification to reduce verbosity, but I don't think it's something that's worth putting effort to standardize.